GSoC/2016/JuliaQuantum-:-Framework-for-QuComput.md

```diff
@@ -0,0 +1,420 @@
+
+**The proposal will realize**:
+
+- a user-friendly type system to help developers add new features or gates in the framework (based on defining gates-like modules as subtype of `AbstractGates`)
+- simulations:
+    + basic quantum circuit
+    + adiabatic quantum computation model
+    + [one-way quantum computation](http://arxiv.org/abs/quant-ph/0602096v1)
+    + [quantum error corrections](https://en.wikipedia.org/wiki/Quantum_error_correction) (eg. [CSS code](https://en.wikipedia.org/wiki/CSS_code)) by quantum circuits and one-way model
+- optimizations:
+    + **quantum circuit optimization** auto-optimize the circuit to an easy-realized form for experiments, eg. three qubit gates could cost a lot more resource than some one or two qubit gates' combinations, so break a three qubit gate into combinations of one and two qubit gates will be necessary. [Qcompiler with CSD method](http://cpc.cs.qub.ac.uk/summaries/AENX_v1_0.html)
+- visualizations:
+    + **quantum circuit visualization** visualize the quantum circuits described by user in this framework by [Compose.jl](https://github.com/dcjones/Compose.jl) or [Luxor.jl](https://github.com/cormullion/Luxor.jl)
+    + **graph states** graph states visualization will be based on [GraphPlot.jl](https://github.com/afternone/GraphPlot.jl)
+
+At the end of the project, JuliaQuantum will be equipped with a framework to simulate quantum computation. Including the features about:
+
+- quantum circuit's simulation and visualization
+- Adiabatic model simulators
+- one-way model's simulation and visualization
+
+The project will make use of existing packages and aims to contribute back, for example:
+
+- `QuArray` will be used to store the states of Qubits
+- QuDynamics.jl will provide adiabatic computational model with more efficient Shordinger equation solvers with good accuracy. 
+
+
+Explanation of nouns:
+
+- **computational models** here means the algorithms that can provide universal computation and is possible to be realized in experiments
+
+
+---
+## Details :
+
+### Structure of Types
+
+![](imag/AbstractGate.png)
+
+
+### Prototype
+I wrote a prototype for my idea with only simple features about quantum circuit and adiabatic quantum simulation module in [Kebab](https://github.com/Roger-luo/Kebab).
+
+### Implementation:
+
+#### Type System
+- **Quantum Gates** `AbstractGates` is a super type for all modules in the structure of this framework. The simulations of a quantum computation model will be construct in a quantum-circuit-like way. The other quantum computation model will be a gate-like module(subtype of `AbstractGates`) in quantum circuit.
+- **Quantum Gate** quantum gates `Gate` includes gate-like modules and single qubit gates like Hadamard gate,Pauli gates...
+- **Graph** graph type will be used to store graph states for one-way model, the type is current considered to use adjacent matrix and use [GraphPlot.jl](https://github.com/afternone/GraphPlot.jl) as dependency to visualize
```

Comment:
yup, thanks. And I think that methods in `AbstractComputationalModel` could generate a certain gate under given parameters, so that users can tweak the parameters more flexible! I'll change the structure to this.